### PR TITLE
feat(syn): Relax Seeds Syntax

### LINKED ISSUE #3727: allow arbitrary expressions in seeds=...
I expected code like this to work:

```rs
const PREFIX: &[u8] = b"prefix";

fn pda_seeds (arg: &Pubkey) -> &[&[u8]] {
    [ PREFIX, arg.as_ref() ]
}

#[derive(Accounts]) struct SomeContext<'info> {
    #[account(mut)]
    signer: Signer<'info>,

    #[account(init, payer = signer, size = 16, seeds = pda_seeds(signer.key()), bump]
    something: Account<'x, ...>,
}
```

However, this constraint:

https://github.com/solana-foundation/anchor/blob/0bdfa3f760635cc83bbda13f9a9d22d1558d1776/lang/syn/src/parser/accounts/constraints.rs#L368

Causes this error:

```
error: expected square brackets
```

Apparently literal brackets are required. Is it realistic to relax that constraint a bit and allow seeds to be abstracted into functions so that we don't have to copy-paste multi-line expressions in attribute macro tags when we want to reuse a PDA

### PR DESCRIPTION
**Pull request by 0xIchigo**
This PR aims to relax the seeds syntax to resolve #3727. This change would allow for both `[ b"prefix".as_ref(), key.as_ref() ]` and `pda_seeds(key)` to be acceptable expressions in `seeds = ...`

### CODE REVIEW

File: lang/syn/src/codegen/accounts/constraints.rs [hunk 1]
```diff
@@ -1145,47 +1171,94 @@ fn generate_constraint_seeds(f: &Field, c: &ConstraintSeedsGroup) -> proc_macro2
...
+                let (__pda_address, __bump) =
+                    Pubkey::find_program_address(__seeds_slice, &#deriving_program_id);
+                __bumps.#name = #bump_store;
             },
-            // Bump target given. Use it.
+
+            // Bump supplied.
             Some(b) => quote! {
+                // Copy into a Vec so we can push the provided bump.
+                let mut __seeds_vec: ::std::vec::Vec<&[u8]> = #expr.to_vec();
```
Comment (Aursen) [inline lang/syn/src/codegen/accounts/constraints.rs]: Why are you using Vec here and not .concat() for example?

File: lang/syn/src/codegen/accounts/constraints.rs [hunk 2]
```diff
@@ -1145,47 +1171,94 @@ fn generate_constraint_seeds(f: &Field, c: &ConstraintSeedsGroup) -> proc_macro2
...
+                        anchor_lang::error::ErrorCode::ConstraintSeeds
+                    ).with_account_name(#name_str)
+                     .with_pubkeys((#name.key(), __pda_address)));
+                }
+            };
+        }
+
+        let expr = match &c.seeds {
+            SeedsExpr::Expr(e) => e, // The user‑supplied expression.
+            _ => unreachable!("handled list case above"),
```
Comment (Aursen) [inline lang/syn/src/codegen/accounts/constraints.rs]: This kind of logic seems weird to me, the enum is to avoid that kind of pattern of code

File: lang/syn/src/codegen/accounts/constraints.rs [hunk 3]
```diff
@@ -1145,47 +1171,94 @@ fn generate_constraint_seeds(f: &Field, c: &ConstraintSeedsGroup) -> proc_macro2
...
-        // Not init here, so do all the checks.
-        let define_pda = match c.bump.as_ref() {
-            // Bump target not given. Find it.
+        if let SeedsExpr::List(list) = &c.seeds {
+            // Work on a *mutable* clone so we can pop / push safely.
+            let mut seeds = list.clone();
+
+            // If the list ends with a trailing comma, `pop()` returns `Pair::Punct`
+            // which we need to turn back into a value before pushing it again.
+            if let Some(pair) = seeds.pop() {
```
Comment (Aursen) [inline lang/syn/src/codegen/accounts/constraints.rs]: if seeds.trailing_punct() {
    seeds.pop_punct();
}

Maybe this is better here

**Review by Aursen — CHANGES_REQUESTED** [review]

File: lang/syn/src/codegen/accounts/constraints.rs [hunk 4]
```diff
@@ -497,71 +497,97 @@ fn generate_constraint_init_group(
...
-                seeds.push_value(pair.into_value());
-            }
+                if let Some(pair) = seeds.pop() {
+                    seeds.push_value(pair.into_value());
+                }
+
+                let maybe_seeds_plus_comma = (!seeds.is_empty()).then(|| quote! { #seeds, });
 
-            let maybe_seeds_plus_comma = (!seeds.is_empty()).then(|| {
-                quote! { #seeds, }
```
Comment (Aursen) [inline lang/syn/src/codegen/accounts/constraints.rs:510]: Maybe don't allow empty seeds? It can simplify the syntax a bit

**Review by Aursen — CHANGES_REQUESTED** [review]
Thank you for the PR

File: lang/syn/src/codegen/accounts/constraints.rs (hunk 4 above)
Comment (0xIchigo) [reply lang/syn/src/codegen/accounts/constraints.rs:510]: Sound good, thanks—will update

File: lang/syn/src/codegen/accounts/constraints.rs (hunk 3 above)
Comment (0xIchigo) [reply lang/syn/src/codegen/accounts/constraints.rs]: Agreed—will implement

File: lang/syn/src/codegen/accounts/constraints.rs (hunk 2 above)
Comment (0xIchigo) [reply lang/syn/src/codegen/accounts/constraints.rs]: Looking back at it, yeah lol that's my bad—will remove

File: lang/syn/src/codegen/accounts/constraints.rs (hunk 1 above)
Comment (0xIchigo) [reply lang/syn/src/codegen/accounts/constraints.rs]: From my understanding, we want to use `Vec` and not `.concat()` because we're appending to the `&[&[u8]]` slice. `.concat()` would flatten the inner slices into a single `Vec<u8>` and change the PDA derivation when we want to add the bump

**Comment by 0xIchigo** [comment]
@jacobcreech @Aursen, my b—all good to review now. I tried going down a few rabbit holes, but it started to add unnecessary complexity and greatly expanded the scope of this PR, so I reverted to a working build and made a few updates

Re `pop_punct()`: Unfortunately, we can't use `pop_punct()` as it's a syn 2.x method and we're on syn 1.0.109, which has `trailing_punct()` but not `pop_punct()`. We could update this in the future—I want to keep this PR as minimal as possible because it already introduces minimal breaking changes

Re: forbidding empty seeds: I'm leaving them allowed for this PR. Deriving a PDA with no explicit seeds (i.e., relying on only the program ID + bumps) is valid, and forbidding globally it would be a breaking change, which blows everything up lol

File: lang/syn/src/lib.rs [hunk 5]
```diff
@@ -873,17 +873,79 @@ pub struct ConstraintInitGroup {
...
 }
 
+/// Seeds can be written as a literal slice (`[ a, b ]`) or any
+/// expression that produces `&[&[u8]]` at run time.
+#[derive(Debug, Clone)]
+pub enum SeedsExpr {
+    /// Example: `[ b"prefix".as_ref(), key.as_ref() ]`
+    List(Punctuated<Expr, Token![,]>),
+    /// Exampl: `pda_seeds(key)`
+    Expr(Box<Expr>),
```
Comment (jamie-osec) [inline lang/syn/src/lib.rs:883]: This doesn't need to be boxed as there's no recursion

File: lang/syn/src/parser/accounts/constraints.rs [hunk 6]
```diff
@@ -364,14 +364,16 @@ pub fn parse_token(stream: ParseStream) -> ParseResult<ConstraintToken> {
...
-                    },
-                ))
+
+                let seeds_expr = if stream.peek(token::Bracket) {
+                    let content;
+                    let _bracket = bracketed!(content in stream);
+                    SeedsExpr::List(content.parse_terminated(Expr::parse)?)
+                } else {
+                    SeedsExpr::Expr(Box::new(stream.parse()?))
+                };
```
Comment (jamie-osec) [inline lang/syn/src/parser/accounts/constraints.rs]: I'd suggest moving this parsing logic to an impl of `Parse`, and stripping any trailing punctuation there:
``
```rs
impl Parse for SeedsExpr {
    fn parse(stream: ParseStream) -> ParseResult<Self> {
        if stream.peek(Bracket) {
            let content;
            bracketed!(content in stream);
            let mut list = content.parse_terminated(Expr::parse)?;
            // Remove any trailing comma - use `pop_punct` when we update to Syn 2.0
            if let Some(pair) = list.pop() {
                list.push_value(pair.into_value());
            }
            Ok(Self::List(list))
        } else {
            Ok(Self::Expr(stream.parse()?))
         }
     }
 }

```

File: lang/syn/src/codegen/accounts/constraints.rs [hunk 7]
```diff
@@ -1145,49 +1171,80 @@ fn generate_constraint_seeds(f: &Field, c: &ConstraintSeedsGroup) -> proc_macro2
...
-                __bumps.#name = #bump;
+        // Build the PDA.
+        let define_pda = match (&c.seeds, &c.bump) {
+            // [list], no bump -> find_program_address + store __bump.
+            (SeedsExpr::List(list), None) => {
+                let mut seeds = list.clone();
+                if let Some(pair) = seeds.pop() {
+                    seeds.push_value(pair.into_value());
+                }
+                let maybe_seeds_plus_comma = (!seeds.is_empty()).then(|| quote! { #seeds, });
```
Comment (jamie-osec) [inline lang/syn/src/codegen/accounts/constraints.rs]: With the suggested change to centralize the comma-removing logic, this can be removed (as can the other instances)

File: lang/syn/src/lib.rs [hunk 8]
```diff
@@ -873,17 +873,79 @@ pub struct ConstraintInitGroup {
...
+/// Allow `quote!{ #seeds }`
+impl quote::ToTokens for SeedsExpr {
+    fn to_tokens(&self, tokens: &mut proc_macro2::TokenStream) {
+        match self {
+            SeedsExpr::List(list) => list.to_tokens(tokens),
+            SeedsExpr::Expr(expr) => expr.to_tokens(tokens),
+        }
+    }
+}
+
```
Comment (jamie-osec) [inline lang/syn/src/lib.rs]: All of these impls can be removed with the suggested changes - I think that treating the two variants homogenously can lead to confusing results (especially the `ToTokens` impl which will either produce a valid expression or an invalid one (i.e. `expr, expr, expr` with no brackets).

File: lang/syn/src/codegen/accounts/constraints.rs [hunk 9]
```diff
@@ -497,71 +497,97 @@ fn generate_constraint_init_group(
...
-            let maybe_seeds_plus_comma = (!seeds.is_empty()).then(|| {
-                quote! { #seeds, }
-            });
-
-            let validate_pda = {
-                // If the bump is provided with init *and target*, then force it to be the
-                // canonical bump.
-                //
-                // Note that for `#[account(init, seeds)]`, find_program_address has already
-                // been run in the init constraint find_pda variable.
```
Comment (jamie-osec) [inline lang/syn/src/codegen/accounts/constraints.rs:518]: A couple of comments were removed here which I think would be useful to keep around

File: lang/tests/seeds_compile.rs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+#[derive(Accounts)]
+pub struct LiteralSeeds<'info> {
+    #[account(
+        // Literal list parsed as `SeedsExpr::List`
+        seeds = [PREFIX, user.key().as_ref()],
+        bump
+    )]
+    pda: Account<'info, Dummy>,
```
Comment (jamie-osec) [inline lang/tests/seeds_compile.rs:34]: I'd suggest explicitly testing for the trailing comma edge case logic
```suggestion
    #[account(
        // Literal list parsed as `SeedsExpr::List`
        seeds = [PREFIX, user.key().as_ref()],
        bump
    )]
    pda: Account<'info, Dummy>,
    #[account(
        // Literal list with a trailing comma parsed as `SeedsExpr::List`
        seeds = [PREFIX, user.key().as_ref(),],
        bump
    )]
```

**Review by jamie-osec — CHANGES_REQUESTED** [review]

File: lang/syn/src/lib.rs (hunk 5 above)
Comment (0xIchigo) [reply lang/syn/src/lib.rs:883]: Clippy flags the enum as `large_enum_variant` because `syn::Expr` is a lot larger than the list variant. Boxing the `Expr` keeps the enum's size small and avoids copying large payloads at the cost of one pointer indirection. We could put a `#[allow(clippy::large_enum_variant)]` on the enum to silence the lint, but keeping it boxed seems cleaner

File: lang/syn/src/lib.rs (hunk 8 above)
Comment (0xIchigo) [reply lang/syn/src/lib.rs]: I'll have to revisit removing this, as it seems like most tests now fail

Comment (0xIchigo) [reply lang/syn/src/lib.rs]: Yeah, it's looking like this will need to stay, or else we'd have to refactor all the failing tests. Curious to hear your thoughts on a potential workaround or if leaving the impls is okay

File: lang/syn/src/lib.rs (hunk 5 above)
Comment (jamie-osec) [reply lang/syn/src/lib.rs:883]: Got it; I thought this might have been a refactoring artifact

File: lang/syn/src/lib.rs (hunk 8 above)
Comment (jamie-osec) [reply lang/syn/src/lib.rs]: Sounds good - maybe doc comments on the needed functions to outline the semantics

**Review by jamie-osec — APPROVED** [review]
LGTM with the changes

File: lang/syn/src/lib.rs (hunk 8 above)
Comment (0xIchigo) [reply lang/syn/src/lib.rs]: Added some comments for clarity

**Review by jacobcreech — APPROVED** [review]
